Common/bTagEffprod.py

```diff
@@ -0,0 +1,81 @@
+    df = df.Define("sample_type", f"static_cast<int>(SampleType::{config[sample_name]['sampleType']})")
+    df = df.Define("period", f"static_cast<int>(Period::{period})")
+    df = df.Define("X_mass", f"static_cast<int>({mass})")
+
+    df = Baseline.CreateRecoP4(df)
+    df = Baseline.DefineGenObjects(df, isData=False, isHH=isHH)
+    dfw = Utilities.DataFrameWrapper(df)
+    dfw.Apply(Baseline.SelectRecoP4, 'nano')
+    dfw.Apply(Baseline.RecoLeptonsSelection)
+    dfw.Apply(Baseline.RecoHttCandidateSelection, config["GLOBAL"])
+    dfw.Apply(Baseline.RecoJetSelection)
+    df = dfw.df
+    pt_bins = Utilities.ListToVector([20,30,40,60,100,150,200,300,1000], "double")
+    eta_bins = Utilities.ListToVector([0,0.6,1.2,2.1,2.4], "double")
+    model = ROOT.RDF.TH2DModel("", "", pt_bins.size()-1, pt_bins.data(), eta_bins.size()-1, eta_bins.data())
+    hists = {}
+    for flav in [ 0, 4, 5]:
+        df = df.Define(f"Jet_eta_flavour{flav}", f"abs(v_ops::eta(Jet_p4[Jet_bCand && Jet_hadronFlavour=={flav}]))")
+        df = df.Define(f"Jet_pt_flavour{flav}", f"v_ops::pt(Jet_p4[Jet_bCand && Jet_hadronFlavour=={flav}])")
+        hist_name = f'jet_pt_eta_{flav}'
+        hists[hist_name] = df.Histo2D(model, f"Jet_pt_flavour{flav}", f"Jet_eta_flavour{flav}")
+        for wp, thr in wpValues.items():
+            df = df.Define(f"Jet_eta_WP{wp.name}_flavour{flav}", f"abs(v_ops::eta(Jet_p4[Jet_bCand && Jet_hadronFlavour=={flav} && Jet_btagDeepFlavB > {thr}]))")
+            df = df.Define(f"Jet_pt_WP{wp.name}_flavour{flav}", f"v_ops::pt(Jet_p4[Jet_bCand && Jet_hadronFlavour=={flav} && Jet_btagDeepFlavB > {thr}])")
+            hist_name = f'jet_pt_eta_{flav}_{wp.name}'
+            hists[hist_name] = df.Histo2D(model, f"Jet_pt_WP{wp.name}_flavour{flav}", f"Jet_eta_WP{wp.name}_flavour{flav}")
+    return hists
+
+if __name__ == "__main__":
+    import argparse
+    import os
+    import yaml
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--configFile', type=str)
+    parser.add_argument('--inFile', type=str)
+    parser.add_argument('--sample', type=str)
+    parser.add_argument('--nEvents', type=int, default=None)
+    parser.add_argument('--evtIds', type=str, default='')
+    parser.add_argument('--outFile', type=str)
+
+    args = parser.parse_args()
+    if os.path.exists(args.outFile):
+        os.remove(args.outFile)
+
+    ROOT.gROOT.ProcessLine(".include "+ os.environ['ANALYSIS_PATH'])
+    ROOT.gROOT.ProcessLine('#include "Common/GenTools.h"')
+    isHH=False
+    with open(args.configFile, 'r') as f:
+        config = yaml.safe_load(f)
+    hists = bTagProdEff(args.inFile, config, args.sample, args.nEvents,args.evtIds)
+    fileToSave = ROOT.TFile(args.outFile, 'RECREATE')
+    for hist_name, hist in hists.items():
+        hist.SetTitle(hist_name)
+        hist.SetName(hist_name)
+        fileToSave.WriteTObject(hist.GetValue(), hist_name)
+    fileToSave.Close()
```

Comment:
run and print validation report checking for empty bins or bins with stat error for eff > thr. Where default thr e.g. 1%.

Reply:
working on it